### PR TITLE
fix: properly check if a member can be nullable

### PR DESCRIPTION
**Pull request by sugmanue**
## Description of changes
Change the way we compute the nullability of members. Before we were only looking at the box trait in the target type and ignoring the trait on the member itself. See for instance the test `can read box trait from member` which has the following model:
```
        namespace com.test
        structure MyStruct {
           @box
           foo: MyFoo
        }
        long MyFoo
```
The code was incorrectly assuming that the foo member was not nullable. This change accounts for this by using the `NullableIndex` which correctly computes whether or not the member or the target type are nullable. 

**NOTE** This is a draft pull request as it depends on the pending release of smithy 1.25.0 scheduled for tomorrow, but sending along to get a head start on the review of the changes.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt
Comment (ianbotsf) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt:44]: **Nit**: There's no need to set this in the `init` block. It can be set directly after the field definition on L26:

```kotlin
private val nullableIndex = NullableIndex(model)
```

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt
Comment (ianbotsf) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt:177]: **Nit**: There's no need to make `targetSymbol` mutable:

```kotlin
val targetSymbol = if (nullableIndex.isMemberNullable(shape, NullableIndex.CheckMode.CLIENT_ZERO_VALUE_V1)) {
    toSymbol(targetShape).toBuilder().boxed().build()
} else {
    toSymbol(targetShape)
}
```

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/StructureGenerator.kt
Comment (ianbotsf) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/StructureGenerator.kt:34]: **Suggestion**: We shouldn't need to reinstantiate the `NullabilityIndex` when it's already present in the `symbolProvider`. I suggest encapsulating the logic to determine if a member is nullable into `KotlinSymbolProvider` and then call that method from this class.

File: smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/core/SymbolProviderTest.kt
Comment (ianbotsf) [inline smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/core/SymbolProviderTest.kt:136]: **Nit**: Missing newline between methods.

**Review by ianbotsf — COMMENTED** [review]
Cannot fully review until Smithy 1.25 is release, this PR is refreshed, and the required checks pass. Leaving some intermediate feedback in the meantime...

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt
Comment (sugmanue) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt:44]: Make sense, I will change it accordingly.

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt
Comment (sugmanue) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt:177]: Make sense, I will change it accordingly.

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/StructureGenerator.kt
Comment (aajtodd) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/StructureGenerator.kt:34]: In most places though I think we rely on the generic `SymbolProvider` rather than the actual `KotlinSymbolProvider` implementation

**Review by aajtodd — COMMENTED** [review]
As Ian said I'll wait for approval until all the checks can run.

This looks ok for now as far as fixing the specific member trait issue but I am wondering if we want to move away from the "boxed" terminology that is pervasive in the current code. ([example](https://github.com/awslabs/smithy-kotlin/blob/v0.12.5/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/model/SymbolExt.kt#L43-L74)). 

I'm assuming we will have follow on PR(s) to handle updating how default values are handled?

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/StructureGenerator.kt
Comment (sugmanue) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/StructureGenerator.kt:34]: I pinged Ian about this yesterday, this change will snowball into a ton of type changes across several files, I don't feel like it's a net-win to be honest, but I already have the changes so I might as well post them for you guys to decide.

Comment (ianbotsf) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/StructureGenerator.kt]: I'm fine with replacing `SymbolProvider` with `KotlinSymbolProvider` as you've posted...despite the number of changes it incurs, I think it's the right change to make.

One alternative worth discussing is retaining the previous type dependencies and just adding an extension method for `isMemberNullable` onto `SymbolProvider`:

```kotlin
fun SymbolProvider.isMemberNullable(member: MemberShape): Boolean = when (this) {
    is KotlinSymbolProvider -> isMemberNullable(member)
    else -> throw NotImplementedError("Cannot determine nullability of $member via $this")
}
```

Comment (sugmanue) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/StructureGenerator.kt:31]: Synced up offline with Ian, we will keep the `NullableIndex` here for the time being.

**Review by aajtodd — APPROVED** [review]
Missing changelog entry. Please fix before merge: https://github.com/awslabs/smithy-kotlin/blob/main/CONTRIBUTING.md#changelog

**Review by ianbotsf — APPROVED** [review]

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt
Comment (ianbotsf) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt]: Ignore this, it's old feedback that wasn't submitted in time for the approval/merge.

**Pull request merged**
